packages/react-query/src/__tests__/prefetch.test.tsx

```diff
@@ -0,0 +1,352 @@
+import { describe, expect, it, vi } from 'vitest'
+import React from 'react'
+import { fireEvent, waitFor } from '@testing-library/react'
+import { ErrorBoundary } from 'react-error-boundary'
+import {
+  QueryCache,
+  usePrefetchInfiniteQuery,
+  usePrefetchQuery,
+  useQueryErrorResetBoundary,
+  useSuspenseInfiniteQuery,
+  useSuspenseQuery,
+} from '..'
+import { createQueryClient, queryKey, renderWithClient, sleep } from './utils'
+
+import type { InfiniteData, UseInfiniteQueryOptions, UseQueryOptions } from '..'
+import type { Mock } from 'vitest'
+
+const generateQueryFn = (data: string) =>
+  vi.fn<any, Promise<string>>().mockImplementation(async () => {
+    await sleep(10)
+
+    return data
+  })
+
+const generateInfiniteQueryOptions = (
+  data: Array<{ data: string; currentPage: number; totalPages: number }>,
+) => {
+  let currentPage = 0
+
+  return {
+    queryFn: vi
+      .fn<any, Promise<(typeof data)[number]>>()
+      .mockImplementation(async () => {
+        const currentPageData = data[currentPage]
+        if (!currentPageData) {
+          throw new Error('No data defined for page ' + currentPage)
+        }
+
+        await sleep(10)
+        currentPage++
+
+        return currentPageData
+      }),
+    initialPageParam: 1,
+    getNextPageParam: (lastPage: (typeof data)[number]) =>
+      lastPage.currentPage === lastPage.totalPages
+        ? undefined
+        : lastPage.currentPage + 1,
+  }
+}
+
+describe('usePrefetchQuery', () => {
+  const queryCache = new QueryCache()
+  const queryClient = createQueryClient({ queryCache })
+
+  function Suspended<TData = unknown>(props: {
+    queryOpts: UseQueryOptions<TData, Error, TData, Array<string>>
+  }) {
+    const state = useSuspenseQuery(props.queryOpts)
+
+    return (
+      <div>
+        <div>data: {String(state.data)}</div>
+      </div>
+    )
+  }
+
+  it('should prefetch query if query state does not exist', async () => {
+    const queryOpts = {
+      queryKey: queryKey(),
+      queryFn: generateQueryFn('Prefetch is nice!'),
+    }
+
+    function App() {
+      usePrefetchQuery(queryOpts)
+
+      return (
+        <React.Suspense fallback="Loading...">
+          <Suspended queryOpts={queryOpts} />
+        </React.Suspense>
+      )
+    }
+
+    const rendered = renderWithClient(queryClient, <App />)
+
+    await waitFor(() => rendered.getByText('data: Prefetch is nice!'))
+    expect(queryOpts.queryFn).toHaveBeenCalledTimes(1)
```

Comment:
Is this testing what it says? It might also be the case that `usePrefetchQuery` didn't fetch, but `useSuspenseQuery` did. Not sure if it necessarily needs fixing, but thought I should point it out.

Reply:
yeah we should probably use a different `queryFn` for prefetching and the component so that we can assert which one ran. @brunolopesr can you add that?

Reply:
@Ephem @TkDodo 

Added a better assertion for this in PR #7614